File: exercise/LeetCode/757. Set Intersection Size At Least Two/Set Intersection Size At Least Two.cpp

#include <iostream>
#include <vector>
#include <algorithm>

using namespace::std;

class Solution {
public:
    int intersectionSizeTwo(vector<vector<int>>& intervals) {
        if (intervals.size() == 0)
            return 0;

        std::sort(intervals.begin(), intervals.end(),
                [](const std::vector<int> a, const std::vector<int>& b){
                    // 按右端升序，右端相同时，按左端降序
                    if (a[1] < b[1]) return true;
                    else if (a[1] == b[1] && a[0] > b[0]) return true;
                    return false;
                });

        // 一个interval最右端的两个点
        int i = intervals[0][1] - 1;
        int j = intervals[0][1];
        int res = 2;
        for (auto &it:intervals)
        {
            // 情况(1)
            if (i < it[0] && it[0] <= j)
            {
                res ++;
                i = j;
                j = it[1];
            }
            // 情况(2)
            else if (j < it[0])
            {
                res += 2;
                i = it[1] - 1;
                j = it[1];
            }
            else
            { }
        }

        return res;
    }
};


int main()
{
    vector<vector<int>> iv;
    vector<int> it(2);
    it[0] = 1; it[1] = 2; iv.push_back(it);
    it[0] = 2; it[1] = 3; iv.push_back(it);
    it[0] = 2; it[1] = 4; iv.push_back(it);
    it[0] = 4; it[1] = 5; iv.push_back(it);

    Solution su;
    int size = su.intersectionSizeTwo(iv);
    std::cout << "Size: " << size << std::endl;

    return 0;
}

File: exercise/LeetCode/757. Set Intersection Size At Least Two/Set Intersection Size At Least Two.py

import os
from functools import cmp_to_key

class Solution:
    def comp(self, a, b):
        """ 比较 a < b : -1 """
        if a[1] < b[1]:
            return -1
        if a[1] == b[1] and a[0] > b[0]:
            return -1
        if a[0] == b[0] and a[1] == b[1]:
            return 0
        return 1

    def intersectionSizeTwo(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        if len(intervals) == 0:
            return 0
        # 两种sort，任选其一
        intervals.sort(key = cmp_to_key(lambda a,b :self.comp(a,b)))
        #intervals.sort(key = lambda x : x[1])
        i = intervals[0][1] - 1
        j = intervals[0][1]
        res = 2
        for it in intervals:
            if i < it[0] and it[0] <= j:
                res += 1
                i = j
                j = it[1]
            elif j < it[0]:
                res += 2
                i = it[1] - 1
                j = it[1]
        return res

if __name__ == "__main__":
    intervals = [[1,3], [1,4], [2,5], [3,5]]
    su = Solution()
    print(su.intersectionSizeTwo(intervals))
